src/Compiler/Checking/import.fs

```diff
@@ -43,48 +45,96 @@ type AssemblyLoader =
     /// Record a root for a [<Generate>] type to help guide static linking & type relocation
     abstract RecordGeneratedTypeRoot : ProviderGeneratedType -> unit
 #endif
-        
+
+[<Struct; NoComparison>]
+type CanCoerce =
+    | CanCoerce
+    | NoCoerce
+
+type [<Struct; NoComparison; CustomEquality>] TTypeCacheKey =
+
+    val ty1: TType
+    val ty2: TType
+    val canCoerce: CanCoerce
+    val tcGlobals: TcGlobals
+
+    private new (ty1, ty2, canCoerce, tcGlobals) =
+        { ty1 = ty1; ty2 = ty2; canCoerce = canCoerce; tcGlobals = tcGlobals }
+
+    static member FromStrippedTypes (ty1, ty2, canCoerce, tcGlobals) =
+        TTypeCacheKey(ty1, ty2, canCoerce, tcGlobals)
+
+    interface System.IEquatable<TTypeCacheKey> with
+        member this.Equals other =
+            if stampEquals this.tcGlobals this.ty1 other.ty1 && stampEquals this.tcGlobals this.ty2 other.ty2 && this.canCoerce = other.canCoerce then
+                true
+            else
+                (
+                    LanguagePrimitives.PhysicalEquality this.ty1 other.ty1
+                    && LanguagePrimitives.PhysicalEquality this.ty2 other.ty2
+                    && this.canCoerce = other.canCoerce
+                )
```

Comment:
```suggestion
            if this.canCoerce <> other.canCoerce then
                false
            elif this.ty1 === other.ty1 && this.ty2 === other.ty2  then
                true 
            else 
                stampEquals this.tcGlobals this.ty1 other.ty1 && stampEquals this.tcGlobals this.ty2 other.ty2                
```


Might need an open for === (physical equality). I wrote that in plain text.
Physical equality of instances is the easiest case and covers also stamp equals, so should come first (if instances are equal, their stamp would be equal too).

Putting canCoerce first (will be branch predicted) so that other conditions can be shorted and do not have to repeat it.